mcp/tool.go

```diff
@@ -97,11 +97,31 @@ func setSchema[T any](sfield **jsonschema.Schema, rfield **jsonschema.Resolved)
 	var err error
 	if *sfield == nil {
 		*sfield, err = jsonschema.For[T](nil)
+		if err != nil {
+			return err
+		}
 	}
+	// Resolve operates with internal state and cannot be called twice on the same
+	// Schema instance. Deep-copy the schema before resolving to avoid mutating the
+	// original (which may be reused when adding the same tool again).
+	cloned, err := func(orig *jsonschema.Schema) (*jsonschema.Schema, error) {
+		if orig == nil {
+			return nil, nil
+		}
+		b, err := json.Marshal(orig)
```

Comment:
I'm not sure if this is the best way to handle this, but @jba will know. Added him as a reviewer.